docs: migration page (#8739)
sveltejs/svelte: documentation/docs/05-misc/04-migrating-v3-to-v4.md

title: Migrating from v3 to v4
---

> This migration guide provides an overview of how to migrate from Svelte version 3 to 4. See the linked PRs for more detailed migration instructions. Use the migration script to migrate some of these automatically: `npx svelte-migrate svelte-4`

Reviewer: "See the linked PRs for more detailed migration instructions" makes it sound like this guide won't be thorough enough. Maybe we can reword?
```suggestion
This migration guide provides an overview of how to migrate from Svelte version 3 to 4. See the linked PRs for more details about each change. Use the migration script to migrate some of these automatically: `npx svelte-migrate svelte-4`
```

> If you're a library author, consider whether to only support Svelte 4 or if it's possible to support Svelte 3, too, after migrating - since most breaking changes don't affect many people, this may be easily possible. Also remember to update your `peerDependencies`.

Reviewer: four commas in the first sentence is too many :laughing: maybe split it into two sentances
```suggestion
If you're a library author, consider whether to only support Svelte 4 or if it's possible to support Svelte 3 too. Since most of the breaking changes don't affect many people, this may be easily possible. Also remember to update the version range in your `peerDependencies`.
```

## Minimum version requirements

- Upgrade to Node 16 or higher. Earlier versions are no longer supported. ([#8566](https://github.com/sveltejs/svelte/issues/8566))
- If you are using Vite, the minimum supported vite-plugin-svelte version is now 2.4.1. If you are a SvelteKit user, upgrade to 1.15.9 or newer to ensure compatibility. ([#8516](https://github.com/sveltejs/svelte/issues/8516))
- If you are using webpack, upgrade to webpack 5 or higher. Earlier versions are no longer supported. ([#8515](https://github.com/sveltejs/svelte/issues/8515))
- If you are using TypeScript, upgrade to TypeScript 5 or higher. Lower versions might still work, but no guarantees are made about that. ([#8488](https://github.com/sveltejs/svelte/issues/8488))

## Browser conditions for bundlers

> Bundlers must now specify the browser condition when building a frontend bundle for the browser. SvelteKit and Vite will handle this automatically for you, for Rollup or webpack you may need to adjust your config. ([#8516](https://github.com/sveltejs/svelte/issues/8516))

Reviewer: ```suggestion
Bundlers must now specify the browser condition when building a frontend bundle for the browser. SvelteKit and Vite will handle this automatically for you. For Rollup or webpack you may need to adjust your config to ensure it matches what is shown in the [`rollup-plugin-svelte`](https://github.com/sveltejs/rollup-plugin-svelte/#usage) and [`svelte-loader`](https://github.com/sveltejs/svelte-loader#usage) documentation. ([#8516](https://github.com/sveltejs/svelte/issues/8516))
```

## Removal of CJS related output

> Svelte no longer supports the CommonJS (CJS) format for compiler output and has also removed the `svelte/register` hook and the CJS runtime version. If you needed those, consider using a bundler like Vite or the full-stack framework SvelteKit instead. ([#8613](https://github.com/sveltejs/svelte/issues/8613))

Reviewer: it would be a lot of work to migrate a non-sveltekit project to sveltekit, so it feels like a harsh migration suggestion. how about this?
```suggestion
Svelte no longer supports the CommonJS (CJS) format for compiler output and has also removed the `svelte/register` hook and the CJS runtime version. If you needed those, consider setting `"type": "module"` in your project and using ESM. This has been set upon creating a new Svelte project for the past couple of years and most of the Svelte ecosystem now expects it. ([#8613](https://github.com/sveltejs/svelte/issues/8613))
```

Author: webpack and rollup set `esm` by default. What I imagined was someone not using a bundler at all, instead handrolling something, which is why they need to cjs output for some reason. You're right that someone could use cjs together with a bundler already. I'm not sure why they would want to do that and what would be needed to migrate, which is why I'd like to keep it rather vague.

## Stricter types for Svelte functions

There are now stricter types for `createEventDispatcher`, `Action`, `ActionReturn`, and `onMount`:

- `createEventDispatcher` now supports specifying that a payload is optional, required, or non-existent, and the call sites are checked accordingly ([#7224](https://github.com/sveltejs/svelte/issues/7224))
- `Action` and `ActionReturn` have a default parameter type of `never` now, which means you need to type the generic if you want to specify that this action receives a parameter. The migration script will migrate this automatically ([#7442](https://github.com/sveltejs/svelte/pull/7442))
- `onMount` now shows a type error if you return a function asynchronously from it, because this is likely a bug in your code where you expect the callback to be called on destroy, which it will only do for synchronously returned functions ([#8136](https://github.com/sveltejs/svelte/issues/8136))

## Custom Elements with Svelte

> The creation of custom elements with Svelte has been overhauled and significantly improved. The `tag` option is deprecated in favor of the new `customElement` option. The migration script will adjust your code automatically. The update timing of properties has changed slightly, too. ([#8457](https://github.com/sveltejs/svelte/issues/8457))

Reviewer: ```suggestion
The creation of custom elements with Svelte has been overhauled and significantly improved. The `tag` option is deprecated in favor of the new `customElement` option. The migration script will adjust your code automatically. The update timing of properties has changed slightly as well. ([#8457](https://github.com/sveltejs/svelte/issues/8457))
```

## SvelteComponentTyped is deprecated

`SvelteComponentTyped` is deprecated, as `SvelteComponent` now has all its typing capabilities. Replace all instances of `SvelteComponentTyped` with `SvelteComponent`. If you have used `SvelteComponent` as the component instance type previously, you may see a somewhat opaque type error now, which is solved by changing `: typeof SvelteComponent` to `: typeof SvelteComponent<any>` (more info in the linked PR). The migration script will do both automatically for you. ([#8512](https://github.com/sveltejs/svelte/issues/8512))

## Transitions are local by default

> Transitions are now local by default to prevent confusion around page navigations. To make them global, add the `|global` modifier. The migration script will do this automatically for you. ([#6686](https://github.com/sveltejs/svelte/issues/6686))

Reviewer: if we have time, it'd be nice to make this optional
